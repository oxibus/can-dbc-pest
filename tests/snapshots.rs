#![cfg(feature = "encoding")]

use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [("tests/shared-test-files/dbc-cantools", "dbc-cantools")];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match DbcParser::parse(Rule::file, &buffer) {
            Ok(pairs) => assert_debug_snapshot!(file_name, pairs),
            Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
        }
    } else {
        panic!("Failed to decode {file_name}.dbc as cp1252");
    }
}
