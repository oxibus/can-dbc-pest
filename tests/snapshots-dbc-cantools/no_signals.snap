---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///\"\n\nBS_:\n\nBU_: Node TestNode\n\n\nBO_ 1024 Message1: 5 TestNode\n\nBO_ 1025 Message2: 0 TestNode\n\n\n\n\n",
            start: 0,
            end: 166,
        },
        inner: [
            Pair {
                rule: line,
                span: Span {
                    str: "VERSION \"HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///\"",
                    start: 0,
                    end: 72,
                },
                inner: [
                    Pair {
                        rule: version,
                        span: Span {
                            str: "VERSION \"HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///\"",
                            start: 0,
                            end: 72,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///\"",
                                    start: 8,
                                    end: 72,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 72,
                    end: 73,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 73,
                    end: 74,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BS_:",
                    start: 74,
                    end: 78,
                },
                inner: [
                    Pair {
                        rule: bit_timing,
                        span: Span {
                            str: "BS_:",
                            start: 74,
                            end: 78,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 78,
                    end: 79,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 79,
                    end: 80,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_: Node TestNode",
                    start: 80,
                    end: 98,
                },
                inner: [
                    Pair {
                        rule: nodes,
                        span: Span {
                            str: "BU_: Node TestNode",
                            start: 80,
                            end: 98,
                        },
                        inner: [
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "Node",
                                    start: 85,
                                    end: 89,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "TestNode",
                                    start: 90,
                                    end: 98,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 98,
                    end: 99,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 99,
                    end: 100,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 100,
                    end: 101,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BO_ 1024 Message1: 5 TestNode",
                    start: 101,
                    end: 130,
                },
                inner: [
                    Pair {
                        rule: message,
                        span: Span {
                            str: "BO_ 1024 Message1: 5 TestNode",
                            start: 101,
                            end: 130,
                        },
                        inner: [
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "1024",
                                    start: 105,
                                    end: 109,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "Message1",
                                    start: 110,
                                    end: 118,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "5",
                                    start: 120,
                                    end: 121,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "TestNode",
                                    start: 122,
                                    end: 130,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 130,
                    end: 131,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 131,
                    end: 132,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BO_ 1025 Message2: 0 TestNode",
                    start: 132,
                    end: 161,
                },
                inner: [
                    Pair {
                        rule: message,
                        span: Span {
                            str: "BO_ 1025 Message2: 0 TestNode",
                            start: 132,
                            end: 161,
                        },
                        inner: [
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "1025",
                                    start: 136,
                                    end: 140,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "Message2",
                                    start: 141,
                                    end: 149,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "0",
                                    start: 151,
                                    end: 152,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "TestNode",
                                    start: 153,
                                    end: 161,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 161,
                    end: 162,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 162,
                    end: 163,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 163,
                    end: 164,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 164,
                    end: 165,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 165,
                    end: 166,
                },
                inner: [],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 166,
                    end: 166,
                },
                inner: [],
            },
        ],
    },
]
